File: src/selector.rs
use log::error;

pub enum Operator {
    Eq,
    Ne,
}

pub struct Selector {
    pub key: String,
    pub value: String,
    pub operator: Operator,
}

impl std::fmt::Display for Selector {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Selector {
                key,
                value,
                operator,
            } => match operator {
                Operator::Eq => write!(f, "{}={}", key, value),
                Operator::Ne => write!(f, "{}!={}", key, value),
            },
        }
    }
}

impl Selector {
    pub fn from(l: &str) -> Selector {
        let not_equal = l.split("!=").collect::<Vec<&str>>();
        let equal_double = l.split("==").collect::<Vec<&str>>();
        let equal_single = l.split('=').collect::<Vec<&str>>();
        let selector = match (not_equal.len(), equal_double.len(), equal_single.len()) {
            (2, _, _) => Selector {
                key: not_equal[0].trim().to_string(),
                value: not_equal[1].trim().to_string(),
                operator: Operator::Ne,
            },
            (_, 2, _) => Selector {
                key: equal_double[0].trim().to_string(),
                value: equal_double[1].trim().to_string(),
                operator: Operator::Eq,
            },
            (_, _, 2) => Selector {
                key: equal_single[0].trim().to_string(),
                value: equal_single[1].trim().to_string(),
                operator: Operator::Eq,
            },
            _ => {
                error!("❌ Invalid label selector format: {}", l);
                panic!("Invalid label selector format");
            }
        };
        if selector.key.is_empty()
            || selector.key.contains('!')
            || selector.key.contains('=')
            || selector.value.is_empty()
            || selector.value.contains('!')
            || selector.value.contains('=')
        {
            error!("❌ Invalid label selector format: {}", l);
            panic!("Invalid label selector format");
        }
        selector
    }
    
}

File: pkg/types/selector.go
package types

import (
	"fmt"
	"log"
	"strings"
)

// Operator represents the comparison operator for selectors
type Operator int

const (
	Eq Operator = iota // Equal
	Ne                 // Not Equal
)

// String returns the string representation of the Operator
func (o Operator) String() string {
	switch o {
	case Eq:
		return "="
	case Ne:
		return "!="
	default:
		return "unknown"
	}
}

// Selector represents a key-value selector with an operator
type Selector struct {
	Key      string
	Value    string
	Operator Operator
}

// String returns the string representation of the Selector
func (s *Selector) String() string {
	return fmt.Sprintf("%s%s%s", s.Key, s.Operator, s.Value)
}

// FromString creates a new Selector from a string representation
func FromString(s string) (*Selector, error) {
	notEqual := strings.Split(s, "!=")
	equalDouble := strings.Split(s, "==")
	equalSingle := strings.Split(s, "=")

	var selector *Selector

	switch {
	case len(notEqual) == 2:
		selector = &Selector{
			Key:      strings.TrimSpace(notEqual[0]),
			Value:    strings.TrimSpace(notEqual[1]),
			Operator: Ne,
		}
	case len(equalDouble) == 2:
		selector = &Selector{
			Key:      strings.TrimSpace(equalDouble[0]),
			Value:    strings.TrimSpace(equalDouble[1]),
			Operator: Eq,
		}
	case len(equalSingle) == 2:
		selector = &Selector{
			Key:      strings.TrimSpace(equalSingle[0]),
			Value:    strings.TrimSpace(equalSingle[1]),
			Operator: Eq,
		}
	default:
		log.Printf("❌ Invalid label selector format: %s", s)
		return nil, fmt.Errorf("invalid label selector format")
	}

	// Validate selector
	if selector.Key == "" ||
		strings.Contains(selector.Key, "!") ||
		strings.Contains(selector.Key, "=") ||
		selector.Value == "" ||
		strings.Contains(selector.Value, "!") ||
		strings.Contains(selector.Value, "=") {
		log.Printf("❌ Invalid label selector format: %s", s)
		return nil, fmt.Errorf("invalid label selector format")
	}

	return selector, nil
}
